apps/extension/src/routes/popup/approval/transaction/use-transaction-views.ts

```diff
@@ -0,0 +1,54 @@
+import { TransactionView } from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/core/transaction/v1alpha1/transaction_pb';
+import { useEffect, useMemo, useState } from 'react';
+import { asPublicTransactionView, asReceiverTransactionView } from '@penumbra-zone/types';
+import { TransactionViewTab } from './types';
+import { useStore } from '../../../../state';
+import { deserializedTransactionViewSelector } from '../../../../state/tx-approval';
+import { isControlledAddress } from './is-controlled-address';
+
+/**
+ * A receiver transaction view is calculated asynchronously, since it needs to
+ * call the async function `isControlledAddress` for each address. So we'll
+ * encapsulate that logic inside its own hook.
+ */
+const useAsReceiver = (transactionView?: TransactionView) => {
+  const [asReceiver, setAsReceiver] = useState<TransactionView | undefined>();
+
+  useEffect(() => {
+    if (transactionView) {
+      asReceiverTransactionView(isControlledAddress)(transactionView)
+        .then(setAsReceiver)
+        .catch(() => {
+          /** no-op */
+        });
+    }
+  }, [transactionView]);
+
+  return asReceiver;
+};
+
+export const useTransactionViews = () => {
```

Comment:
If it weren't for the async call, I'd say to create a synchronous selector out of this whole hook.  But given the call, I think we should use react-query here as the single hook (aka no useEffect) and a synchronous function to handle all the logic.

Reply:
Other thoughts:

I really try to keep async stuff out of hooks. I've found this is the number one most challenging aspect of React and quite difficult to reason about. React-Query has helped, but I think `Data Loaders` are just better.

How about we make the selection a query parameter (or route), load the data in a react-router data loader, and then the component can expect that is available. This will allow us to avoid stuffing async logic in our UI code.

Reply:
Ah, great idea. I haven't used the data loader pattern much yet, but it's a good one. Will rework this to use that.